### PR TITLE
Create a chat app

### LINKED ISSUE #33: Missing Chat App
Creating database related operations for a chat app.

This includes:
Setting up the chat app.
Adding the Message model.
Adding migration files for initializing the data model.
Adding data model functions and model manager classes.

### PR DESCRIPTION
**Pull request by shacharlevy**
- Setup the chat app.
- Add the Message model to models.py file (the fields that are related to the other models are commented out).
- Add filters functions to the manager(commented out).
- Override the Message save function to make sure that always one of the student_course_id and group_id fields will be null (also commented out).
- Register Message in admin.py in order to display the data in admin page.
- Create Migration files for initializing the data model.
- Add the chat app to installed apps in the HighTeach setting.py.

fixes #33

### CODE REVIEW

File: chat/models.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+#     def get_student_course_chat(self, student_course_id):
+#         return self.filter(student_course_id=student_course_id).order_by('date_time')
+
+
+class Message(models.Model):
+    message_id = models.BigAutoField(primary_key=True)
+    # sender_id = models.ForeignKey("User", on_delete=models.CASCADE)
+    # student_course_id = models.ForeignKey("Student_Course", on_delete=models.CASCADE)
+    # group_id = models.ForeignKey("Group", on_delete=models.CASCADE)
+    date_time = models.DateTimeField(default=timezone.now,  blank=False)
```
Comment (shirazyomtov) [inline chat/models.py:18]: I suggest setting the date_time field to be filled automatically according to the time the message was sent.

File: chat/models.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# class ChatManager(models.Manager):
+#     def get_group_chat(self, group_id):
+#         return self.filter(group_id=group_id).order_by('date_time')
+#
+#     def get_student_course_chat(self, student_course_id):
+#         return self.filter(student_course_id=student_course_id).order_by('date_time')
+
+
+class Message(models.Model):
+    message_id = models.BigAutoField(primary_key=True)
```
Comment (idan1356) [inline chat/models.py:14]: I think `AutoField` is sufficient for the ID instead of `BigAutoField`

File: chat/models.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    # sender_id = models.ForeignKey("User", on_delete=models.CASCADE)
+    # student_course_id = models.ForeignKey("Student_Course", on_delete=models.CASCADE)
+    # group_id = models.ForeignKey("Group", on_delete=models.CASCADE)
+    date_time = models.DateTimeField(default=timezone.now,  blank=False)
+    message = models.CharField(max_length=255,  blank=False)
+    # objects = ChatManager()
+
+    # def save(self, *args, **kwargs):
```
Comment (idan1356) [inline chat/models.py:22]: I've been reading about it and its usually not recommended to do the validation logic inside the save method. 

There is a specific method called `clean` that gets invoked  by Django validation-system and that's where the validation logic is supposed to be. 

however, the method `clean`  is only automatically invoked when the instance is created by `ModelForm` or `ModelSerializer`. meaning that when creating an instance programmatically, it is needed to call `clean` manually (although I believe chat messages will be always created via a form anyway)

I hope I understood this feature correctly

File: chat/models.py (hunk 2 above)
Comment (idan1356) [inline chat/models.py:14]: I think `AutoField` is sufficient for the ID instead of `BigAutoField`

File: chat/models.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    #     if self.student_course_id is None and self.group_id is None:
+    #         raise Exception("You can't leave both student_course_id and group_id fields as null")
```
Comment (idan1356) [inline chat/models.py:24]: According to what I've seen the exception that needs to be raised is `ValidationError`

**Review by idan1356 — CHANGES_REQUESTED** [review]
Great! added a few comments

File: chat/models.py (hunk 1 above)
Comment (shacharlevy) [reply chat/models.py]: I changed it :)

File: chat/models.py (hunk 2 above)
Comment (shacharlevy) [reply chat/models.py]: I wanted to be consistent with the other PRs.
In addition, in the case of chat where there are lots of instant messages, it is possible to surpass the autofield limit (in theory) compared to other models. 
What do you think?

File: chat/models.py (hunk 4 above)
Comment (shacharlevy) [reply chat/models.py]: Fixed :)
I read the documentation about the implementation of custom model validations and I moved the validation to the clean method.
The clean method is not invoked on save() or create() by default, so I overriden the save method of the model and invoked the full_clean().

File: chat/models.py (hunk 2 above)
Comment (idan1356) [reply chat/models.py]: Sounds reasonable! No need to change then.

**Review by shirazyomtov — DISMISSED** [review]
LGTM

**Review by idan1356 — DISMISSED** [review]
LGTM!

**Review by natifridman — COMMENTED** [review]
Looks good. Do you want to merge it with the comments or wait for other models?

**Comment by shacharlevy** [comment]
> Looks good. Do you want to merge it with the comments or wait for other models?

I will wait for the other models, thank you :)

**Comment by galitf** [comment]
Please resolve the conflict

**Review by shirazyomtov — APPROVED** [review]
LGTM

**Review by Davidco94 — APPROVED** [review]
LGTM!

**Comment by galitf** [comment]
Please resolve the conflict

**Comment by shacharlevy** [comment]
> Please resolve the conflict

Done :)

**Review by idan1356 — APPROVED** [review]
LGTM

**Comment by natifridman** [comment]
> I will wait for the other models, thank you :)

@shacharlevy are we still waiting for other models?

**Review by natifridman — APPROVED** [review]

**Pull request merged**